### PR TITLE
Adorable badger uat delivery planning creation

### PR DESCRIPTION
**Pull request by metas-rc**
https://github.com/metasfresh/metasfresh/issues/13942

### CODE REVIEW

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningCreateRequest.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	IncotermsId incotermsId;
+
+	@Nullable
+	ShipmentScheduleId shipmentScheduleId;
+
+	@Nullable
+	ReceiptScheduleId receiptScheduleId;
+
+	@Nullable
+	LocalDateAndOrgId plannedLoadingDate;
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningCreateRequest.java:101]: why not using Instant and avoid using OrgDAO in DeliveryPlanningRepository

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningRepository.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+import org.springframework.stereotype.Repository;
+
+import static org.adempiere.model.InterfaceWrapperHelper.load;
+import static org.adempiere.model.InterfaceWrapperHelper.newInstance;
+import static org.adempiere.model.InterfaceWrapperHelper.save;
+
+@Repository
+public class DeliveryPlanningRepository
+{
+	private final IOrgDAO orgDAO = Services.get(IOrgDAO.class);
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningRepository.java:53]: a repo shall not use another repo/dao (if possible)

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningRepository.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+	private final IQueryBL queryBL = Services.get(IQueryBL.class);
+
+	public I_M_Delivery_Planning getById(@NonNull final DeliveryPlanningId deliveryPlanningId)
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningRepository.java:57]: why is this public? i mean, why exposing the record outside repo?

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningRepository.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+		deliveryPlanningRecord.setQtyTotalOpen(qtyTotalOpen.toBigDecimal());
+		deliveryPlanningRecord.setActualDeliveredQty( actualDeliveredQty.toBigDecimal());
+		deliveryPlanningRecord.setActualLoadQty(actualLoadQty.toBigDecimal());
+
+		deliveryPlanningRecord.setForwarder(request.getForwarder());
+		deliveryPlanningRecord.setWayBillNo(request.getWayBillNo());
+		deliveryPlanningRecord.setReleaseNo(request.getReleaseNo());
+		deliveryPlanningRecord.setTransportDetails(request.getTransportDetails());
+
+		deliveryPlanningRecord.setIsActive(request.isActive);
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningRepository.java:107]: why not using the method?

and mainly why are we creating inactive records? what's the requirement?

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningService.java [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	private final IProductBL productBL = Services.get(IProductBL.class);
+
+	private final DeliveryPlanningRepository deliveryPlanningRepository;
+
+	public DeliveryPlanningService(@NonNull final DeliveryPlanningRepository deliveryPlanningRepository)
+	{
+		this.deliveryPlanningRepository = deliveryPlanningRepository;
+	}
+
+	public boolean autoCreateEnabled(@NonNull final ClientAndOrgId clientAndOrgId)
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningService.java:79]: why not prefix it with `is`?

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningService.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+	{
+		return sysConfigBL.getBooleanValue(SYSCONFIG_M_Delivery_Planning_CreateAutomatically, false, clientAndOrgId);
+	}
+
+	private final IShipmentScheduleBL shipmentScheduleBL = Services.get(IShipmentScheduleBL.class);
+
+	private final IReceiptScheduleBL receiptScheduleBL = Services.get(IReceiptScheduleBL.class);
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningService.java:86]: pls move all services on top

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningRepository.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+		deliveryPlanningRecord.setM_Delivery_Planning_Type(DeliveryPlanningType.toCodeOrNull(request.getDeliveryPlanningType()));
+
+		deliveryPlanningRecord.setBatch(request.getBatch());
+		deliveryPlanningRecord.setC_OriginCountry_ID(CountryId.toRepoId(request.getOriginCountryId()));
+		deliveryPlanningRecord.setC_DestinationCountry_ID(CountryId.toRepoId(request.getDestinationCountryId()));
+
+		save(deliveryPlanningRecord);
+	}
+
+	public boolean otherDeliveryPlanningsExistForOrderLine(@NonNull final OrderLineId orderLineId, @NonNull final DeliveryPlanningId excludeDeliveryPlanningId)
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningRepository.java:121]: why not prefix with `is`?

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningService.java [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+
+		final boolean otherDeliveryPlanningsExistForOrderLine = deliveryPlanningRepository.otherDeliveryPlanningsExistForOrderLine(orderLineId, DeliveryPlanningId.ofRepoId(deliveryPlanning.getM_Delivery_Planning_ID()));
+
+		if (!otherDeliveryPlanningsExistForOrderLine)
+		{
+			throw new AdempiereException(MSG_M_Delivery_Planning_AtLeastOnePerOrderLine);
+		}
+	}
+
+	private DeliveryPlanningCreateRequest createRequest(@NonNull final I_M_Delivery_Planning deliveryPlanningRecord)
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningService.java:117]: i think u can make this method static

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/process/M_DeliveryPlanning_CreateAdditionalLines.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+{
+	private final DeliveryPlanningService deliveryPlanningService = SpringContextHolder.instance.getBean(DeliveryPlanningService.class);
+
+	private final String PARAM_AdditionalLines = "AdditionalLines";
+	@Param(parameterName = PARAM_AdditionalLines)
+	private int additionalLines;
+	@Override
+	protected String doIt() throws Exception
+	{
+		Check.assumeGreaterThanZero(additionalLines, PARAM_AdditionalLines);
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/process/M_DeliveryPlanning_CreateAdditionalLines.java:46]: fine to have this checking here but IMHO it makes sense to have it in `createAdditionalDeliveryPlannings` too

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/api/impl/ReceiptScheduleBL.java [hunk 10]
```diff
@@ -732,4 +764,67 @@ public void setBPartnerAddress(final String address)
...
+				.receiptScheduleId(ReceiptScheduleId.ofRepoId(receiptScheduleRecord.getM_ReceiptSchedule_ID()))
+				.deliveryPlanningType(DeliveryPlanningType.Incoming)
+				.orderId(orderId)
+				.orderLineId(orderLineId)
+				.warehouseId(WarehouseId.ofRepoId(receiptScheduleRecord.getM_Warehouse_ID()))
+				.productId(ProductId.ofRepoId(receiptScheduleRecord.getM_Product_ID()))
+				.partnerId(BPartnerId.ofRepoId(receiptScheduleRecord.getC_BPartner_ID()))
+				.bPartnerLocationId(BPartnerLocationId.ofRepoId(receiptScheduleRecord.getC_BPartner_ID(), receiptScheduleRecord.getC_BPartner_Location_ID()))
+				.sectionCodeId(SectionCodeId.ofRepoIdOrNull(receiptScheduleRecord.getM_SectionCode_ID()))
+				.isActive(receiptScheduleRecord.isActive())
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/api/impl/ReceiptScheduleBL.java:800]: this does not make any sense IMHO

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/modelvalidator/M_ReceiptSchedule.java [hunk 11]
```diff
@@ -37,13 +40,21 @@
 import org.adempiere.warehouse.validationrule.FilterWarehouseByDocTypeValidationRule;
 import org.compiere.model.I_C_OrderLine;
 import org.compiere.model.ModelValidator;
+import org.springframework.stereotype.Component;
 
 import java.math.BigDecimal;
 
 @Interceptor(I_M_ReceiptSchedule.class)
+@Component
```
Comment (teosarca) [inline backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/modelvalidator/M_ReceiptSchedule.java:48]: i think its fine, but i just want to make sure u checked somehow to not get this interceptor registered twice, so code will be called twice...

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningRepository.java (hunk 4 above)
Comment (metas-rc) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningRepository.java:107]: Got rd of it. It was not needed

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/modelvalidator/M_ReceiptSchedule.java (hunk 11 above)
Comment (metas-rc) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/inoutcandidate/modelvalidator/M_ReceiptSchedule.java:48]: It was called here https://github.com/metasfresh/metasfresh/pull/14010/files#diff-4834f924890ccfab5af8331a22402c710a90fc503ba1a83ca54a0b3b1a63b34a . But I deleted it.

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningService.java (hunk 8 above)
Comment (metas-rc) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/deliveryplanning/DeliveryPlanningService.java:117]: I can't, because of the UOM

**Review by teosarca — APPROVED** [review]

**Pull request merged**